File: iosApp/iosApp/Utils/AuthManager.swift
//
//  AuthManager.swift
//  iosApp
//
//  Created by Dmitry Kuklin on 08.07.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared
import OSLog

protocol AuthManagerProtocol {
    func checkAuth() -> AuthState
    func checkOnboardingState() -> Bool
    func setOnboardingState()
    
    func checkValetAvailability(name: String) -> Bool
    func register(with name: String)
}

class AuthManager: AuthManagerProtocol, LoggerManager {
    let logger: Logger = Logger()
    
    //TODO: Check AuthManagerApi address not to create multiple instance. and avoid race conditions
    var authManagerApi = AuthManagerApi(factory: DriverFactory())
    
    func checkAuth() -> AuthState {
        log()
        return authManagerApi.getAuthStatus()
    }
    
    func checkOnboardingState() -> Bool {
        Logger().info("AuthManagerProtocol -- checkOnboardingState")
        return authManagerApi.getOnboardingStatus()
    }
    
    func setOnboardingState() {
        return authManagerApi.setOnboardingCompletion(isCompleted: true)
    }

    func checkValetAvailability(name: String) -> Bool {
        return name != "Dima" && !name.isEmpty
    }
    
    func register(with name: String) {
        authManagerApi.setAuthStatus(name: name)
    }
}

File: androidApp/src/main/java/utils/AuthManager.kt
package utils

import android.content.Context
import com.example.metasecret.AuthManagerApi
import com.example.metasecret.Greeting
import com.example.metasecret.database.DriverFactory
import com.jetbrains.handson.kmm.shared.cache.AuthState
import javax.inject.Inject

class AuthManager @Inject constructor() {
    fun checkValetAvailability(name: String): Boolean {
        //TODO: Read from DB
        return true
    }

    fun checkAuth(context: Context): AuthState {
        return AuthManagerApi(factory = DriverFactory(context)).getAuthStatus()
    }

    fun register(name: String, context: Context) {
        AuthManagerApi(factory = DriverFactory(context)).setAuthStatus(name = name)
    }

    fun checkOnboardingState(context: Context): Boolean {
        return AuthManagerApi(factory = DriverFactory(context)).getOnboardingStatus()
    }

    fun setOnboardingState(isCompleted: Boolean, context: Context) {
        return AuthManagerApi(factory = DriverFactory(context)).setOnboardingCompletion(isCompleted)
    }
}